### ISSUE TITLE
FI_HMEM support: ofi_cudaMemcpy() using default stream can cause dead lock.

### ISSUE THREAD

**Issue by wzamazon** [ofiwg/libfabric#6124]
I encountered this issue when I was testing the FI_HMEM support of the EFA provider with NCCL, and I believe this problem will apply to any provider that use ```ofi_cudaMemcpy()``` to copy data to GPU during message transfer.

The core issue here is that ```ofi_cudaMemcpy()``` use ```cudaMemcpy()``` to deliver data, which is on default stream. Meanwhile, for efficiency, some application/library might launch a kernel that wait for libfabric to deliver data and process the data. If the kernel is also on default stream, there will be a dead lock.

Take NCCL as an example, 

When NCCL uses cuda buffer directly, it will allocate bounce buffer using GPU memory.
When a NCCL routine is called with default stream, it will launch a cuda kernel that is on that stream. 
That cuda kernel is waiting for data to be copied into bounce buffer, then copy the data from bounce buffer to target buffer and other GPU.
If an libfabric provider use ```ofi_cudaMemcpy```to deliver data,  because that function uses default stream, there will be  a dead lock.

Note the problem does not limit to NCCL. In principal, any message transfer application/library (MPI for example) that want to launch a kernel on a default stream that wait for libfabric to deliver data will have this issue. And they might want to do that for efficiency.

**Comment by wzamazon** [comment]
To reproduce the problem, you need:

1. An application that uses NCCL. I used the example C program on NCCL website:
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/examples.html
The last one is what I used.

2. NCCL: I tried 2.4.8 and 2.5.6 both have the issue

3. A NCCL network plugin. that support cuda pointer

4. A libfabric that support FI_HMEM capability.

Note, one might need to adjust the environment variable  NCCL_NET_GDR_LEVEL to make NCCL use cuda buffer directly.

**Comment by rajachan** [comment]
Cc: @iziemba. Did you run into this with any of your testing? Using the default stream for the copies seem to be an issue and we might have to change the ofi_copy_cuda_* implementation.

**Comment by shefty** [comment]
It sounds like the API may need to be updated to give the app control over where copies occur for a given MR.

**Comment by iziemba** [comment]
I have not seen deadlock issues. So is the following happening?

An application launches a kernel on the default stream. At some point, the kernel is polling on device memory waiting for the completion of an async receive. During the execution of the kernel, the host is calling fi_cq_read() to progress data transfers (or a thread is used). To progress, a call into cudaMemcpy is needed. Since the default stream is used, the cudaMemcpy won't be executed until the kernel completes. But, since the kernel is waiting for the cudaMemcpy to progress, it cannot complete. Does this sound correct?

**Comment by wzamazon** [comment]
@iziemba Yes, this is exactly what happened in this example.

**Comment by iziemba** [comment]
Could we have libfabric allocate its own CUDA stream?

**Comment by wzamazon** [comment]
I tried the approach that have libfabric create a new separate CUDA stream and use cudaMemcpyAsync with it.

This approach solved the above issue but encountered another issue. The following is an example code to demonstrate the new issue:
```
cudaMalloc(&sendBuff, size*sizeof(float));
cudaMalloc(&recvBuff, size*sizeof(float));
cudaMallocHost(&hostBuff, size*sizeof(float));
ncclAllreduce(sendoff, recvbuff, size, ncclFloat, ncclSum, comm, 0);
cudaMemcpy(hostbuff, recvbuff, size, cudaMemcopyDeviceToHost);
```
Note, the main difference is the ```cudaMemcpy``` we added after ```ncclAllreduce```.
The problem was that because ncclAllreduce() is non-blocking, it will launch kernel and create a thread that call fi_cq_read(), then exit. 

The following call to cudaMemcpy() will apply a pthread mutex lock on cuda resources, then wait for the nccl kernel to finish. With the mutex lock on cuda resources, the alternative stream approach will not work either because cudaMemcpyAsync will also be blocked by the lock.

**Comment by jdinan** [comment]
NCCL plugins are not allowed to make CUDA calls because of the potential for deadlock. There are a few ways to avoid CUDA. You can use GPU direct RDMA. You can use [GDRCOPY](https://github.com/NVIDIA/gdrcopy) to perform the copy from the CPU side. This routes the copy through PCIe BAR1, avoiding CUDA calls. Alternatively, you may be able to have NCCL place data on the CPU before the plugin gets called (I'm less familiar with how to do this, but can help if needed).

Agree with @shefty, allowing the application layer to control when CUDA calls can be made safely will be helpful to work around limitations like this.

**Comment by shefty** [comment]
I don't know the details, but this sounds like a flaw in either the hw or sw design.

**Comment by iziemba** [comment]
Sounds like libfabric is not meeting a NCCL requirement due to:

1. CUDA streams blocking until previous commands complete
2. NCCL kernels polling device memory to progress collectives
3. Libfabric using CUDA calls to move data to device memory

As @jdinan pointed out, using GPU direct RDMA or GDRCOPY could solve the issue. For example, the verbs provider could be configured to exclusively use GPU direct RDMA for this case. But, once RXM is layered over verbs, RXM bounce buffers may be used thus GDRCOPY may be required. In addition, if we want to ever support FI_HMEM over a non-RDMA provider (tcp for example), then GDRCOPY may be required.

I assume NCCL must have a progression thread which is calling fi_cq_read(). @jdinan Would building libfabric with CUDA_API_PER_THREAD_DEFAULT_STREAM enabled make a difference at all?

**Comment by rajachan** [comment]
Thanks Jim. GDRCOPY might solve the problem of copying data into the device when a kernel is expecting it, but it is not going to be ideal for performance when copying from device to host, is that correct?  If it is, we still need a more general solution that would work for libfabric.

In the NCCL case, staging through the CPU (`NCCL_PTR_HOST` only, in NCCL terms), is doable but that takes away the bandwidth gains to be had when using a provider's native FI_HMEM support. Also, we would have no control over how other runtimes that use FI_HMEM uses the streams, so we're going to fix our copy routines or our APIs anyway.

**Comment by shefty** [comment]
It sounds like the issue is that the NCCL plug-in is non-compliant.  Maybe there's something that can be done with that implementation that would avoid the providers needing to perform a memcpy?  I agree that we should fix-up our copy routines.  Maybe we need more configuration information passed down by the app, such as the correct stream to use, in order to avoid deadlocks?

**Comment by rajachan** [comment]
In the case @wzamazon outlined, the EFA provider was initiating the copies on behalf of the NCCL plug-in, so we can fix this in the EFA provider by avoiding the eager transfers involving copies when using CUDA buffers.

The plug-in just calls into libfabric (FI_TAGGED) for all sizes and leaves it up to a provider to pick the right transfer protocol based on size. (It is a pretty thin plugin really, <1500 SLOC)

**Comment by wzamazon** [comment]
We have introduced infrastructure to use gdrcopy, also the EFA provider implemented to use a local RDMA read to copy data to GPU buffer.
